### PR TITLE
Task 11988+90+92: Getting Started Screen: Create UI + Handle Event + Fetch API

### PR DESCRIPTION
**Pull request by vanpd-0859**
### Issue Link :link:

https://edu-redmine.sun-asterisk.vn/issues/11985
### Goals :soccer:

Create UI
https://marvelapp.com/be90ba4/screen/56503605
- Title name
- Genre buttons
- Next button

Handle Event:
- Click to genre buttons to select favorite genres
- Click Next button to go to Home Screen

Fetch data and display to UI:
- Genre list

### Implementation Details :construction:

https://gfycat.com/OffbeatMealyFinch

### CODE REVIEW

**Comment by vanpd-0859** [comment]
ready

File: MovieApp/Data/Source/Remote/API/Middleware/CustomRequestAdapter.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+import Foundation
+import Alamofire
+
+final class CustomRequestAdapter: RequestAdapter {
+    func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        var urlRequest = urlRequest
+        if let accessToken = UserDefaults.standard.string(forKey: APIs.apiKey) {
```
Comment (tienpx3010) [inline MovieApp/Data/Source/Remote/API/Middleware/CustomRequestAdapter.swift]: Thừa

File: MovieApp/Data/Source/Remote/API/Request/GenreListRequest.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+import ObjectMapper
+import Alamofire
+
+final class GenreListRequest: BaseRequest {
+    required init() {
+        let body: [String: Any]  = [:]
+        super.init(url: URLs.genreListApi, requestType: .get, body: body)
```
Comment (tienpx3010) [inline MovieApp/Data/Source/Remote/API/Request/GenreListRequest.swift]:
```suggestion
        super.init(url: URLs.genreListApi, requestType: .get, body: [:])
```

File: MovieApp/Scenes/GettingStarted/Cell/GenreItemViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,11 @@
+//
+//  GenreCellViewModel.swift
+//  MovieApp
+//
+//  Created by Phan Dinh Van on 5/14/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+struct GenreItemViewModel {
+    let genreTitle = ""
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/Cell/GenreItemViewModel.swift]:
```suggestion
    let genreTitle
```

File: MovieApp/Scenes/GettingStarted/Cell/GettingStartedGenreCell.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        genreTitleLabel.do {
+            $0.textColor = .white
+        }
+    }
+    
+    func bind(viewModel: GenreItemViewModel) {
+        genreTitleLabel.text = viewModel.genreTitle
+    }
+    
+    func updateCell(genre: Genre?) {
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/Cell/GettingStartedGenreCell.swift:35]: Sử dụng bind

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        let navigator = DefaultGettingStartedNavigator(viewController: self)
+        viewModel = GettingStartedViewModel(genreRepository: genreRepository, navigator: navigator)
+        let input = GettingStartedViewModel.Input(loadTrigger: Driver.just(()),
+                                                  selectTrigger: genreListCollectionView.rx.itemSelected.asDriver(),
+                                                  doneButtonTrigger: doneButton.rx.tap.asDriver())
+        let output = viewModel.transform(input: input)
+        output.genres
+            .drive(genreListCollectionView.rx.items) { collectionView, index, genre in
+                let indexPath = IndexPath(row: index, section: 0)
+                let cell: GettingStartedGenreCell = collectionView.dequeueReusableCell(for: indexPath)
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift:56]:
```suggestion
                let cell: GettingStartedGenreCell = collectionView.dequeueReusableCell(for: indexPath).then {
                        $0.bind(viewModel: GenreItemViewModel(genre: genre))
                }
```

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                cell.updateCell(genre: genre)
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift]: Bỏ

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                return cell
+            }
+            .disposed(by: disposeBag)
+        
+        output.selected
+            .asObservable()
+            .subscribe(onNext: { indexPath in
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift]: drive

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                self.genreListCollectionView.deselectItem(at: indexPath, animated: false)
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift]: Tạo binder

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        output.indicator
+            .drive(SVProgressHUD.rx.isAnimating)
+            .disposed(by: disposeBag)
+        
+        output.error
+            .drive(errorBinding)
+            .disposed(by: disposeBag)
+    }
+    
+    var errorBinding: Binder<Error> {
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift]: Bỏ vào extension của ViewController

File: MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+//
+//  Created by Phan Dinh Van on 5/13/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+import UIKit
+
+final class GettingStartedViewController: UIViewController, AlertViewController {
+    @IBOutlet private weak var genreListCollectionView: UICollectionView!
+    @IBOutlet weak var doneButton: UIButton!
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewController.swift]:
```suggestion
    @IBOutlet private weak var doneButton: UIButton!
```

File: MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+//  Created by Phan Dinh Van on 5/13/19.
+//  Copyright © 2019 nguyen.nam.khanh. All rights reserved.
+//
+
+import RealmSwift
+
+final class GettingStartedViewModel: ViewModelType {
+    private let genreRepository: GenreReponsitory
+    private let navigator: GettingStartedNavigator
+    private var selectedGenres = Variable<[Genre]>([])
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift]: Sử dụng behavior relay thay vì variable

File: MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            .asDriver(onErrorJustReturn: IndexPath())
+        
+        let isDoneButtonEnabled = selectedGenres.asDriver()
+            .map { selectedGenres in
+                return !selectedGenres.isEmpty
+            }
+        
+        let doneButton = input.doneButtonTrigger
+            .asObservable()
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift]: thừa

File: MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            .flatMap { _ -> Observable<Void> in
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift]:
```suggestion
            .flatMap { _ -> Driver<Void> in
```

File: MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+                do {
+                    let realm = try Realm()
+                    try realm.write {
+                        realm.add(self.selectedGenres.value, update: false)
+                    }
+                    self.navigator.toMain()
+                    return Observable.just(())
+                } catch {
+                    return Observable.error(Errors.cantInitRealm)
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/GettingStartedViewModel.swift:68]: Sử dụng errorTracker để catch error

**Comment by vanpd-0859** [comment]
ready

File: MovieApp/Scenes/GettingStarted/Cell/GenreItemViewModel.swift [hunk 15]
```diff
@@ -7,5 +7,9 @@
 //
 
 struct GenreItemViewModel {
-    let genreTitle = ""
+    let genreTitle: String
```
Comment (tienpx3010) [inline MovieApp/Scenes/GettingStarted/Cell/GenreItemViewModel.swift]: let genre: Genre

var title: String {
      return genre.name
}

File: MovieApp/Extension/AlertViewController+.swift [hunk 16]
```diff
@@ -48,4 +48,18 @@ extension AlertViewController where Self: UIViewController {
     func showErrorAlert(message: String?) {
         showAlertView(title: "Error", message: message, cancelButton: "OK")
     }
+    
+    var errorBinding: Binder<Error> {
+        return Binder(self, binding: { (viewController, error) in
+            if let error = error as? BaseError {
+                viewController.showErrorAlert(message: error.errorMessage)
+            }
+            else if let error = error as? Errors {
```
Comment (tienpx3010) [inline MovieApp/Extension/AlertViewController+.swift]: Errors kế thừa từ BaseError là đc

**Comment by vanpd-0859** [comment]
ready

File: MovieApp/Scenes/GettingStarted/Cell/GettingStartedGenreCell.swift (hunk 4 above)
Comment (tienpx3010) [reply MovieApp/Scenes/GettingStarted/Cell/GettingStartedGenreCell.swift:35]: Xóa đi em nhé

**Review by tienpx3010 — APPROVED** [review]

**Pull request merged**